### PR TITLE
WIP: Optimize wrapping algorithm (e.g. in Rajagopal model)

### PR DESCRIPTION
**Pull request by adamkewley**
This is a placeholder PR for wrapping optimizations, which will be the next thing I will focus on after shipping the `CacheVariable` stuff.

Because the CacheVariable work is probably going to be held up with the usual upstreaming + merging + review process, I'll start to take measurements and elucidate exactly what this change is going to entail.

Rough TODOs:

- [ ] try to gain basic understanding of the problem space
- [x] measure performance of Rajagopal model simulation
- [x] use profile to identify any clear performance bottlenecks
- [ ] implement any additional testing (if tests are missing)
- [ ] implement possible perf improvements

### CODE REVIEW

**Comment by adamkewley** [comment]
## Initial Profiling

`valgrind --tool=callgrind`ing the Rajagopal forward dynamic yields a virtualized measure of performance (i.e. whatever `valgrind` thinks is an accurate reproduction of how my Ryzen would execute the code). This is a useful way of measuring performance because it usually gives  accurate estimates of cycle counts, L1 misses etc. Here's an example call graph from the profile that skips ~20 % of computation (other parts of the simulation):

rajagopal-graph

This is very different from the profile for `ToyDropLanding`, which is what my previous optimization work focused on. In `ToyDropLanding`, there were no obvious bottlenecks because performance hits were "scattered" all over OpenSim as a side-effect of its architecture (lookup functions, cache variables, etc.). By contrast, this call graph clearly indicates that ~50 % of computation time is probably spent in `OpenSim::WrapObject::wrapPathSegment`, which is a relatively low-level and isolated function that performs a bunch of maths to figure out a wrap.

Before I take a (deliciously tempting) dive into the low-level code to start optimizing this, I think the first step is for me to understand what this code is ultimately trying to achieve. Is there something at a higher level that will essentially achieve the same goal, but with less work? Can fewer iterations be performed? Can the algorithm be tweaked a little at a high-level? 

I have no idea--this wrapping work is new for me--but it feels like I should probably try to get a basic understanding before I start tweaking with the memory layout of the matrices or something extremely low-level that'll only change perf a little.

I already have a comparison script from previous perf work, so I at least have a basic benchmarking method (i.e. "how many times faster is it post-change?"), which I'll use moving forward.

**Comment by adamkewley** [comment]
## High-level Overview

#### tl;dr

- Assuming the rest of the OpenSim "stack" (time stepping, integration, force calculation) is sane, which is a *productive*, rather than *correct*, assumption for me to make (it would be time-inefficient for me to inspect the entire stack), I'm going to assume that all relevant perf work should happen in `OpenSim::GeometryPath::applyWrapObjects`
- This alg computes a "wrapped" path. The most hardest part of this work is going to be testing (within some tolerance) that the algorithm generates a *correct* path, rather than that it's *fast*, because defining correctness (e.g. knowing what we're willing to accept) is going to define performance. 
- Once we have a correctness metric (if there isn't one already), then we can grind perf

---

#### Notes

- The profile is of a forward-dynamics (FD) simulation. In an FD simulation, the next step in time is computed (that is, the model's evolution as time moves forward is computed). Essentially, model state is provided, time is stepped forward, new model state computed by interpolating across the timestep.
- One part of computing the next step is computing the forces of all actuators in the model, because those forces will be applied to the skeleton. In the Rajagopal model, ~75-80 % of computation time is spent on computing forces, so this PR focuses on that.
- Muscles are modelled as `OpenSim::PathActuator`s, which are actuators that operate along a path (duh), and a path is defined as a sequence of connected points. 
- The Rajagopal model contains "wrapped" paths, which is where designers can specify a path (A -- B -- C) that must "wrap" around geometry (e.g. a user-defined cylinder). The wrapping feature allows designers to specify that a muscle "wraps around the pelvis", making the simulation responsible for ensuring that the wrapping is "correct" (e.g. the muscle doesn't clip through the cylinder).
- To compute the force of a path actuator, you first need to compute the speed at which it's currently lengthening (e.g. a shortening speed implies a momentum, and force, associated with the shortening).
  - **Side note**: once the lengthening speed is known, you can compute the actuation via various algorithms. Concretely, this profile mostly uses Millard2012 equilibrium muscles, which describe an alg that can compute the muscle force from the lengthening speed (and a bunch of other params). Computing the lengthening speed (overall) is ~50 % of CPU. Computing forces from that lengthening speed is ~15 % CPU, which is why I'm going to ignore the force-computation step and focus entirely on the lengthening step.
- If the path is concretely known, computing the lengthening speed of a path is a matter of adding up the lengthening speed of each segment in the path, which is cheap (~2 %). 
- With wrapping surfaces, the path is not concretely known because it has to be computed dynamically (i.e. the user-defined points and geometry are "resolved" at runtime by "wrapping" the path around the geometry). 
- This algorithm, housed under `OpenSim::GeometryPath::applyWrapObjects` is ~50  % CPU for this benchmark.

**Comment by adamkewley** [comment]
So I've reviewed the top-level of `applyWrapObjects` and it all seems mostly sane in there. I refactored it as I read it because some parts of the code were not clear without reading and re-reading it. These refactors are compiling + passing tests etc (although whether the tests verify the algorithm or not is still something I need to look into).

**Comment by adamkewley** [comment]
Note: that refactor isn't going anywhere near central. After witnessing how long it takes them to merge anything I'm basically going to park any and all non-major refactoring + improvement work.

Instead, I'm going to refocus this on actually writing some comparative simulations with data + profiles, so that we can finally establish the effect of the wrapping surfaces on (e.g.) the integrator's step size (something that I've been avoiding for weeks now because I know that it will take a while to figure out how to write + wire up appropriate models) and begin writing some demo code for using simbody's analogue to wrapping.

**Comment by adamkewley** [comment]
So after adding some tracing code into GeometryPath.cpp (e.g. [gist](https://gist.github.com/adamkewley/1bccb7e059c8605e246a77ce6a28d641)), I can make it print out an average number of iterations which, for the RajagopalModel, seems to be ~2:

```bash
../relwithdebinfo-install/bin/opensim-cmd run-tool RajagopalModel/default_Setup_ForwardTool.xml 
# ...(loads of stuff)...
y[203] = /forceset/vasmed_l/fiber_length

[info] Integrating from 0.0 to 1.0.
[info] Printing results of investigation default to forward
num_iterations = 2.09617
```

it seems like `RajagopalModel` is averaging 2 iterations or so, which isn't particularly bad (this is only an average that does not take outliers etc. into account, though).

**Comment by adamkewley** [comment]
This is roughly in line with what a valgrind trace is showing:

| function | calls |
| ---------- | ------ |
| OpenSim::PathActuator::computeForce | 347 119 |
| OpenSim::GeometryPath::getLengtheningSpeed | 1 735 600 |
| OpenSim::GeometryPath::computeLengtheningSpeed | 1 735 599 |
| OpenSim::GeometryPath::getCurrentPath | 347 119 | 
| OpenSim::GeometryPath::computePath | ~1 200 000 |
| OpenSim::GeometryPath::applyWrap | 347 119 |
| OpenSim::WrapObject::wrapPath | 460 265 | 

Although the trace indicates that `wrapPath` is called, on average, 1.3 times each time that `applyWrap` is called, which indicates that many of those two-iteration calculations exit before getting to `wrapPathSegment`. There's a few premature-exit points (e.g. bounds checks etc.) that could do that.

**Comment by adamkewley** [comment]
So it doesn't look like--at least for the RajagopalModel--the algorithm is having to do many iterations to converge. Hopefully (ignoring bugs), this is a sign that the algorithm can compute an answer in a small number of iterations (rather than having to churn a lot). 

This makes the next point of interest the implementation, but I'm still deliberately avoiding the low-level for now, until we have some basic tests that I can run to ensure everything is (superficially, at least) working fine. Once I have some tests, an obvious candidate for optimization is the matrix maths (e.g. everything in `WrapMath.cpp`) because it feels like that code could be optimized (and the linear algebra is sucking up *a lot* of cycles).
